narwhal/types/src/primary.rs

```diff
@@ -151,6 +227,14 @@ impl BatchAPI for BatchV1 {
     fn metadata_mut(&mut self) -> &mut Metadata {
         &mut self.metadata
     }
+
+    fn versioned_metadata(&self) -> &VersionedMetadata {
+        unimplemented!("BatchV1 does not have a VersionedMetadata field");
```

Comment:
I do not know if there is a better way around it. Maybe @mystenmark  has some ideas? But the two ideas I was tossing up between was just returning a default version of the field that does not exist for the corresponding version or panic. Went with panic to prevent situations where we would mistakingly be using the wrong version. 

With that being said now that we are passing in the protocol config we will just guard all code paths with a feature flag and there will be no mixture of versioned types allowed anyways. Once the network upgrades to the next protocol version we would not be speaking to anyone that is on a lower protocol version. We will also guard against byzantine validators sending us the wrong batch version by checking to see what version we receive from other validators and returning an error if it is unsupported